src/store.rs

```diff
@@ -1053,323 +1179,231 @@ impl StoredObject {
+    fn from(_: Failed) -> Self {
+        UpdateError::Fatal
     }
 }
 
 
-//------------ TreeNames -----------------------------------------------------
-
-/// The names for the manifest and object trees for a given repository.
-///
-/// This type exists so we are guaranteed to always use the same names.
-struct TreeNames<'a> {
-    /// The RRDP notification URI if this is for an RRDP repository.
-    repository_name: &'a [u8],
-}
-
-impl TreeNames<'static> {
-    /// The prefix for manifest trees.
-    const MANIFEST_TREE_PREFIX: &'static [u8] = b"store:manifests:";
+//============ Helper Functions ==============================================
 
-    /// The prefix for object trees.
-    const OBJECT_TREE_PREFIX: &'static [u8] = b"store:objects:";
+/// Creates an IO error of kind other with the given string.
+fn io_err_other<T>(
+    err: impl Into<Box<dyn error::Error + Send + Sync>>
+) -> Result<T, io::Error> {
+    Err(io::Error::new(io::ErrorKind::Other, err))
 }
 
-impl<'a> TreeNames<'a> {
-    /// Returns the tree names for a RRDP repository.
-    pub fn rrdp(rpki_notify: &'a uri::Https) -> Self {
-        TreeNames { repository_name: rpki_notify.as_ref() }
-    }
-
-    /// Returns the tree names for the rsync repository.
-    pub fn rsync() -> Self {
-        TreeNames { repository_name: b"rsync" }
-    }
+/// Recursively iterate over all files in a directory tree.
+fn for_each_file(
+    base: &Path,
+    mut op: impl FnMut(&Path) -> Result<(), Failed>
+) -> Result<(), Failed> {
+    fn recurse(
+        base: &Path,
+        op: &mut impl FnMut(&Path) -> Result<(), Failed>
+    ) -> Result<(), Failed> {
+        let dir = match fs::read_dir(base) {
+            Ok(dir) => dir,
+            Err(err) if err.kind() == io::ErrorKind::NotFound => {
+                return Ok(())
+            }
+            Err(err) => {
+                error!(
+                    "Failed to read directory {}: {}",
+                    base.display(), err
+                );
+                return Err(Failed)
+            }
+        };
 
-    /// Returns the tree names for a given manifest tree name.
-    pub fn from_manifest_tree_name(
-        tree_name: &'a [u8]
-    ) -> Option<Self> {
-        if tree_name.starts_with(TreeNames::MANIFEST_TREE_PREFIX) {
-            Some(TreeNames { repository_name: &tree_name[16..] })
-        }
-        else {
-            None
+        for entry in dir {
+            let entry = match entry {
+                Ok(entry) => entry,
+                Err(err) => {
+                    error!(
+                        "Failed to read directory {}: {}",
+                        base.display(), err
+                    );
+                    return Err(Failed)
+                }
+            };
+            let ftype = match entry.file_type() {
+                Ok(ftype) => ftype,
+                Err(err) => {
+                    error!(
+                        "Failed to read directory {}: {}",
```

Comment:
not a directory per se, I think?

Reply:
At this stage we are still reading the directory (content), not yet the object it points to. So, I think the message is correct?